### PR TITLE
[NOD-401] Add CLI Wallet demo

### PR DESCRIPTION
**Pull request by svarogg**

### CODE REVIEW

File: database/database.go [hunk 1]
```diff
@@ -53,6 +54,7 @@ func Connect(cfg *config.KasparovFlags) error {
 		return err
 	}
 
+	db.LogMode(true)
```
Comment (someone235) [inline database/database.go]: Is it necessary? If so, can you make detailed log in another log level than `error`?

File: kasparovd/controllers/transaction.go [hunk 2]
```diff
@@ -149,12 +148,12 @@ func areTxsInBlock(blockID uint64, txIDs []uint64) (map[uint64]bool, error) {
 
 // GetUTXOsByAddressHandler searches for all UTXOs that belong to a certain address.
 func GetUTXOsByAddressHandler(address string) (interface{}, error) {
-	_, err := util.DecodeAddress(address, config.ActiveConfig().ActiveNetParams.Prefix)
-	if err != nil {
-		return nil, httpserverutils.NewHandlerErrorWithCustomClientMessage(http.StatusUnprocessableEntity,
-			errors.Wrap(err, "error decoding address"),
-			"The given address is not a well-formatted P2PKH or P2SH address.")
-	}
+	//_, err := util.DecodeAddress(address, config.ActiveConfig().ActiveNetParams.Prefix)
```
Comment (someone235) [inline kasparovd/controllers/transaction.go]: Why?

File: demos/wallet/balance.go [hunk 3]
```diff
@@ -0,0 +1,30 @@
+package main
```
Comment (someone235) [inline demos/wallet/balance.go]: why demos and not examples?

File: demos/wallet/balance.go [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+	"github.com/kaspanet/kaspad/util"
+)
+
+func balance(conf *balanceConfig) error {
+	utxos, err := getUTXOs(conf.APIAddress, conf.Address)
+	if err != nil {
+		return err
+	}
+
+	var availableBalance, pendingBalance uint64 = 0, 0
```
Comment (someone235) [inline demos/wallet/balance.go]: We usually don't explicitly initialize to default value, no?

File: demos/wallet/config.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+	createSubCmd  = "create"
+	balanceSubCmd = "balance"
+	sendSubCmd    = "send"
+)
+
+type createConfig struct {
+}
+
+type balanceConfig struct {
+	APIAddress string `long:"api-address" short:"a" description:"Address of API-Server" required:"true"`
```
Comment (someone235) [inline demos/wallet/config.go]: API -> kasparov (in the whole code)

File: demos/wallet/config.go [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+func parseCommandLine() (subCommand string, config interface{}) {
+	cfg := &struct{}{}
+	parser := flags.NewParser(cfg, flags.PrintErrors|flags.HelpFlag)
+
+	createConf := &createConfig{}
+	parser.AddCommand(createSubCmd, "Creates a new wallet",
+		"Creates a new wallet and prints it's private key as well as addresses to all networks", createConf)
+
+	balanceConf := &balanceConfig{}
+	parser.AddCommand(balanceSubCmd, "Shows the balance for a given address",
```
Comment (someone235) [inline demos/wallet/config.go]: for -> of

File: demos/wallet/config.go [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+		"Shows the balance for a given address", balanceConf)
+
+	sendConf := &sendConfig{}
+	parser.AddCommand(sendSubCmd, "Sends a transaction to given address",
+		"Sends a transaction to given address", sendConf)
+
+	_, err := parser.Parse()
+
+	if err != nil {
+		if err, ok := err.(*flags.Error); ok && err.Type == flags.ErrHelp {
```
Comment (someone235) [inline demos/wallet/config.go]: Why don't you print the error?

File: demos/wallet/create.go [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+import (
+	"fmt"
+	"os"
+
+	"github.com/kaspanet/kaspad/dagconfig"
+	"github.com/kaspanet/kaspad/ecc"
+	"github.com/kaspanet/kaspad/util"
+)
+
+func create(*createConfig) {
```
Comment (someone235) [inline demos/wallet/create.go]: `*createConfig` is not used, and `create` is not part of any interface, so this argument is redundant

File: demos/wallet/send.go [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+}
+
+func sendTx(conf *sendConfig, msgTx *wire.MsgTx) error {
+	txBuffer := bytes.NewBuffer(make([]byte, 0, msgTx.SerializeSize()))
+	if err := msgTx.KaspaEncode(txBuffer, 0); err != nil {
+		return err
+	}
+
+	txHex := make([]byte, hex.EncodedLen(txBuffer.Len()))
+	hex.Encode(txHex, txBuffer.Bytes())
```
Comment (someone235) [inline demos/wallet/send.go]: You can use `hex.EncodeToString`

**Review by someone235 — CHANGES_REQUESTED** [review]

File: demos/wallet/config.go (hunk 7 above)
Comment (svarogg) [reply demos/wallet/config.go]: Because the cli parser thingy prints it by itself.

File: demos/wallet/create.go (hunk 8 above)
Comment (svarogg) [reply demos/wallet/create.go]: I use it to conserve the convention where every command function is `commandName(*commandConfig)`

File: demos/wallet/config.go [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+	parser.AddCommand(createSubCmd, "Creates a new wallet",
+		"Creates a private key and 3 public addresses, one for each of MainNet, TestNet and DevNet", createConf)
+
+	balanceConf := &balanceConfig{}
+	parser.AddCommand(balanceSubCmd, "Shows the balance of a public address",
+		"Shows the balance for a public address in Kaspa", balanceConf)
+
+	sendConf := &sendConfig{}
+	parser.AddCommand(sendSubCmd, "Send a Kaspa transaction to a public address",
+		"Send a Kaspa transaction to a public address", sendConf)
```
Comment (stasatdaglabs) [inline demos/wallet/config.go]: The other two are written in present simple whereas this one is in command form. Please use either of them for all three.

File: demos/wallet/common.go [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+	"net/http"
+
+	"github.com/kaspanet/kasparov/kasparovd/apimodels"
+	"github.com/pkg/errors"
+)
+
+func getUTXOs(apiAddress string, address string) ([]*apimodels.TransactionOutputResponse, error) {
+	response, err := http.Get(fmt.Sprintf("%s/utxos/address/%s", apiAddress, address))
+	if err != nil {
+		return nil, errors.Wrap(err, "Error getting utxos from Kasparov server")
```
Comment (stasatdaglabs) [inline demos/wallet/common.go]: Please capitalize UTXOs everywhere a user can see.

File: demos/wallet/send.go [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+
+	utxos, err := getUTXOs(conf.KasparovAddress, fromAddress.String())
+	if err != nil {
+		return err
+	}
+
+	sendAmountSompi := uint64(conf.SendAmount * util.SompiPerKaspa)
+	totalToSend := sendAmountSompi + feeSompis
+
+	selectedUTXOs, change, err := selectUTXOs(utxos, totalToSend)
```
Comment (stasatdaglabs) [inline demos/wallet/send.go]: I'd prefer it if you renamed change to changeSompi, for clarity.

File: demos/wallet/send.go [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+		totalValue += utxo.Value
+
+		if totalValue >= totalToSpend {
+			break
+		}
+	}
+
+	if totalValue < totalToSpend {
+		return nil, 0, errors.Errorf("Insufficient funds for send: %f required, while only %f available",
+			float64(totalToSpend)/util.SompiPerKaspa, float64(totalValue)/util.SompiPerKaspa)
```
Comment (stasatdaglabs) [inline demos/wallet/send.go]: This is likely to print a weird float-y number like 2.1023445130464. Please format the floats to only have a few digits after the decimal point. I recommend two digits.

**Review by stasatdaglabs — CHANGES_REQUESTED** [review]

File: demos/wallet/create.go (hunk 8 above)
Comment (someone235) [reply demos/wallet/create.go]: but still....

File: demos/wallet/send.go [hunk 14]
```diff
@@ -145,11 +145,10 @@ func sendTx(conf *sendConfig, msgTx *wire.MsgTx) error {
 		return err
 	}
 
-	txHex := make([]byte, hex.EncodedLen(txBuffer.Len()))
-	hex.Encode(txHex, txBuffer.Bytes())
+	txHex := hex.EncodeToString(txBuffer.Bytes())
```
Comment (someone235) [inline demos/wallet/send.go]: redundant new line

**Review by someone235 — CHANGES_REQUESTED** [review]

File: demos/wallet/send.go (hunk 13 above)
Comment (svarogg) [reply demos/wallet/send.go]: It's going to be exactly up to 8 decimal points, and this is also the number you want to get.

**Review by someone235 — DISMISSED** [review]

**Review by someone235 — APPROVED** [review]

**Review by stasatdaglabs — APPROVED** [review]

**Pull request merged**